src/Paket.Core/Dependencies/PackageResolver.fs

```diff
@@ -641,35 +643,131 @@ type private Stage =
+        { StartWork : unit -> System.Threading.Tasks.Task
+          mutable Priority : WorkPriority }
+
+type WorkHandle<'a> = private { Handle : Guid; Work : RequestWork; TaskSource : TaskCompletionSource<'a> }
+and ResolverRequestQueue =
+    private { DynamicQueue : System.Collections.Concurrent.ConcurrentDictionary<Guid, RequestWork> }
+module ResolverRequestQueue =
+    open System.Threading
+
+    let Create() = { DynamicQueue = new System.Collections.Concurrent.ConcurrentDictionary<Guid, RequestWork>(); }
+    let addWork prio (f: unit -> System.Threading.Tasks.Task<'a>) ({ DynamicQueue = queue } as q) =
+        let tcs = new TaskCompletionSource<_>()
+        let work =
+            { StartWork = (fun () ->
+                let t =
+                    try
+                        f()
+                    with e -> 
+                        //Task.FromException (e)
+                        let tcs = new TaskCompletionSource<_>()
+                        tcs.SetException e
+                        tcs.Task
+                
+                t.ContinueWith(fun (t:System.Threading.Tasks.Task<'a>) -> 
+                    if t.IsCanceled then
+                        tcs.SetException(new TaskCanceledException(t))
+                    elif t.IsFaulted then
+                        tcs.SetException(t.Exception)
+                    else tcs.SetResult (t.Result)))
+              Priority = prio }
+        let handle = Guid.NewGuid()
+        let res = queue.TryAdd(handle, work)
+        assert res
+        { Handle = handle; Work = work; TaskSource = tcs }
+    let rec private getNext ({ DynamicQueue = queue } as d) =
+        if queue.Count = 0 then None
+        else
+            let min =
+                // cannot minBy as the sequence might be empty by now.
+                queue |> Seq.fold (fun currentMin kv ->
+                    match currentMin with
+                    | None -> Some (kv.Key, kv.Value)
+                    | Some (minKey, min) when kv.Value.Priority < min.Priority -> Some (kv.Key, kv.Value)
+                    | min -> min) None
+            match min with
+            | Some (minKey, min) ->
+                match queue.TryRemove(minKey) with
+                | true, min ->
+                    Some min
+                | _ ->
+                    getNext d
+            | None -> getNext d
+    let startProcessing (cts:CancellationToken) ({ DynamicQueue = queue } as q) =
+        async {
+            while not cts.IsCancellationRequested do
+                match getNext q with
+                | None -> do! Async.Sleep 10
+                | Some work ->
+                    do! work.StartWork().ContinueWith(fun (t:System.Threading.Tasks.Task) ->()) |> Async.AwaitTask
+        }
+        |> Async.StartAsTask
+
+type WorkHandle<'a> with
+    member x.Reprioritize prio =
+        let { Work = work } = x
+        work.Priority <- prio
+    member x.Task =
+        let { TaskSource = task } = x
+        task.Task
```

Comment:
@tpetricek Do you have an idea how I can achieve something like the above.
I basically want a queue but it's critical to re-prioritize some work of it on-demand.

Basically the paket resolver is queuing up work before it actually needs it, but when it needs the info the resolver blocks and therefore the work should be started immediately if not started already. So instead of re-prioritizing another way would be to take the work out of the queue and start it on the resolver directly.

Re-prioritizing generally has the nice property that we can further improve that in the future (ie the more insights the resolver has the more it changes the priority of some work, or even cancels it).

I think the above is a good start, but has some things I don't like:
 - It needs to use sleep so in worst case everything is delayed all the time (I think that's the reason why it's slower than before)
 - When the queue is close to empty a lot of workers go into this bad looking fold (minBy with throw errors...)

If you know other people who might be able to help please ping them. I have ideas but most of them will result in several hundred lines of code and are very hard to get right :/